mlflow/projects/databricks.py

```diff
@@ -27,6 +47,112 @@ def _jobs_runs_submit(req_body_json):
+                                                      os.path.basename(dbfs_fuse_tar_uri)))
+    project_dir = os.path.join(DB_PROJECTS_BASE, tar_hash)
+    mlflow_run_arr = list(map(shlex_quote, ["mlflow", "run", project_dir, "--new-dir",
+                                            "--entry-point", entry_point]))
+    if parameters is not None:
+        for key, value in parameters.items():
+            mlflow_run_arr.extend(["-P", "%s=%s" % (key, value)])
+    mlflow_run_cmd = " ".join(mlflow_run_arr)
+    shell_command = textwrap.dedent("""
+    export PATH=$PATH:$DB_HOME/python/bin:/$DB_HOME/conda/bin &&
+    mlflow --version &&
+    # Make local directories in the container into which to copy/extract the tarred project
+    mkdir -p {0} && mkdir -p {1} &&
+    # Rsync from DBFS FUSE to avoid copying archive into local filesystem if it already exists
+    rsync -a -v --ignore-existing {2} {0} &&
+    # Extract project into a temporary directory. We don't extract directly into the desired
+    # directory as tar extraction isn't guaranteed to be atomic
+    cd $(mktemp -d) &&
+    tar -xzvf {3} &&
+    # Atomically move the extracted project into the desired directory
+    mv -T {4} {5} &&
+    {6}
+    """.format(DB_TARFILE_BASE, DB_PROJECTS_BASE, dbfs_fuse_tar_uri, container_tar_path,
+               DB_TARFILE_ARCHIVE_NAME, project_dir, mlflow_run_cmd))
+    return ["bash", "-c", shell_command]
+
+
+def _check_databricks_cli_configured():
+    cfg_file = os.path.join(os.path.expanduser("~"), ".databrickscfg")
+    try:
+        process.exec_cmd(["databricks", "--version"])
+    except process.ShellCommandException:
+        raise ExecutionException(
+            "Could not find Databricks CLI on PATH. Please install and configure the Databricks "
+            "CLI as described in https://github.com/databricks/databricks-cli")
+    if not os.path.exists(cfg_file):
+        raise ExecutionException("Could not find profile for Databricks CLI in %s. Make sure the "
+                                 "the Databricks CLI is installed and that credentials have been "
+                                 "configured as described in "
+                                 "https://github.com/databricks/databricks-cli" % cfg_file)
+
+
+def _upload_to_dbfs(src_path, dbfs_uri):
+    """
+    Uploads the file at `src_path` to the specified DBFS URI within the Databricks workspace
+    corresponding to the default Databricks CLI profile.
+    """
+    process.exec_cmd(cmd=["databricks", "fs", "cp", src_path, dbfs_uri])
+
+
+def _dbfs_path_exists(dbfs_uri):
+    """
+    Returns True if the passed-in path exists in DBFS for the workspace corresponding to the
+    default Databricks CLI profile.
+    """
+    dbfs_path = _parse_dbfs_uri_path(dbfs_uri)
+    res = rest_utils.databricks_api_request(
+        endpoint="dbfs/get-status", method="GET", params={"path": dbfs_path},
+        verify_success=False)
```

Comment:
This API is a bit wonky. I think we should always return the response code, and maybe only retry on 500-level and connection-related errors, not 400-level errors. Otherwise we risk the client being flaky in the case of temporary server unavailability. What do you think?